### ISSUE TITLE
Releases are very unfriendly for new people

### ISSUE THREAD

**Issue by g00nix** [moby/moby#39302]
The SDK is very unfriendly for beginners.

Apart from the fact every program imports `docker/docker` (including `docker-ce` and `kubernetes`) and that redirects to `moby/moby` on the website, the default version you get from go is `v1.13.1` and people just seem to import specific commits. The releases can not be seen on `moby/moby` but can be found on `docker/engine` however you can not import specific version from there, since other dependencies will crash. 

Even the documentation on the website does not work. This is from https://docs.docker.com/develop/sdk/examples/:
```
	statusCh, errCh := cli.ContainerWait(ctx, resp.ID, container.WaitConditionNotRunning)
	select {
	case err := <-errCh:
		if err != nil {
			panic(err)
		}
	case <-statusCh:
	}
```
`cli.ContainerWait` does not return chans, it returns `(int64, error)` on v1.13.1, and the website does not say how to get other versions of the SDK. Even the documentation on `godoc.org` is wrong since it tells you to get `import "github.com/docker/docker/client"` and that will get you `v1.13.1`, which does not return `(<-chan container.ContainerWaitOKBody, <-chan error)` as specified in the documentation.

Are there any plans to offer some proper documentation and releases so that people can actually use the SDK, without having to dig from git?

**Comment by kolyshkin** [comment]
@gun1x which go tool do you use to get `github.com/docker/docker`?

**Comment by kolyshkin** [comment]
in case you're using go modules, you need to add something like
```
replace github.com/docker/docker => github.com/docker/engine v19.03.0
```
to `go.mod` file.

Let us know if it works, and we'll update the documentation

**Comment by gun1x** [comment]
> @gun1x which go tool do you use to get `github.com/docker/docker`?

I use go version `go1.12.6 linux/amd64` at the moment. I just use the default `go build`, `go run .` and so on.

> in case you're using go modules, you need to add something like
> 
> ```
> replace github.com/docker/docker => github.com/docker/engine v19.03.0
> ```
> 
> to `go.mod` file.
> 
> Let us know if it works, and we'll update the documentation


Let's consider this `main.go` (the one from your documentation, where I changed `docker/docker` with `docker/engine`):
```
package main

import (
	"fmt"

	"github.com/docker/engine/api/types"
	"github.com/docker/engine/client"
	"golang.org/x/net/context"
)

func main() {
	cli, err := client.NewEnvClient()
	if err != nil {
		panic(err)
	}

	containers, err := cli.ContainerList(context.Background(), types.ContainerListOptions{})
	if err != nil {
		panic(err)
	}

	for _, container := range containers {
		fmt.Println(container.ID)
	}
}
```


If you do this without changing version, it will just go for `docker/docker v1.13.1` so the code will just fail with:
```
./main.go:17:87: cannot use "github.com/docker/engine/api/types".ContainerListOptions literal (type "github.com/docker/engine/api/types".ContainerListOptions) as type "github.com/docker/docker/api/types".ContainerListOptions in argument to cli.ContainerList
```
So I tried to change `go.mod` (I don't know if this is what you suggested or not):

```
module testinggo

go 1.12

require (
	github.com/docker/distribution v2.7.1+incompatible // indirect
	github.com/docker/engine v19.3.0 
	github.com/docker/go-connections v0.4.0 // indirect
	github.com/docker/go-units v0.4.0 // indirect
	github.com/opencontainers/go-digest v1.0.0-rc1 // indirect
	github.com/pkg/errors v0.8.1 // indirect
	golang.org/x/net v0.0.0-20190620200207-3b0461eec859
)
```

And the error I got is:
```
 $ go run .
go: finding github.com/docker/engine v19.03.0
go: errors parsing go.mod:
/home/gunix/tmp/go_test/go.mod:7: invalid module version "v19.03.0": unknown revision v19.03.0
```

**Comment by cpuguy83** [comment]
@gun1x This is the incorrect syntax.

Go (and go mods) works via imports.
The import path is github.com/docker/docker, but you want to fetch it from github.com/docker/engine

So you do
```
require (
	github.com/docker/distribution v2.7.1+incompatible // indirect
	github.com/docker/docker v1.13.1
	github.com/docker/go-connections v0.4.0 // indirect
	github.com/docker/go-units v0.4.0 // indirect
	github.com/opencontainers/go-digest v1.0.0-rc1 // indirect
	github.com/pkg/errors v0.8.1 // indirect
	golang.org/x/net v0.0.0-20190620200207-3b0461eec859
)

replace github.com/docker/docker => github.com/docker/engine <commit/tag>
```

**Comment by cpuguy83** [comment]
We discussed this on the maintainers call today, and basically see it as a couple of problems:

1. This repo is not semver
2. The client package had breaking changes sometime between 1.13 and now.
3. Confusion around docker/docker and docker/engine repos (and import paths)

**Comment by gun1x** [comment]
> We discussed this on the maintainers call today, and basically see it as a couple of problems:
> 
> 1. This repo is not semver
> 2. The client package had breaking changes sometime between 1.13 and now.
> 3. Confusion around docker/docker and docker/engine repos (and import paths)

Thank you for the interest. I think there are a lot of people that would like to see some improvements around these points.

Please let me know if there is anything else that I could do.

**Comment by thepudds** [comment]
@cpuguy83 Is this the canonical issue that tracks the need to update how modules-based consumers should import docker?

Or is there a better issue to follow (e.g., maybe https://github.com/moby/moby/issues/38507#issuecomment-456225148, or another issue)?

**Comment by cpuguy83** [comment]
I started experimenting with a new client here: https://github.com/cpuguy83/go-docker/
Do not use it for anything real as it will definitely have breaking changes (I have several queued up as I'm messing with a few things).

Goal is to create something that's easy to semver, isn't necessarily tied to docker releases, and is simpler to use.

**Comment by thepudds** [comment]
Hi all, I wanted to see see if I could bounce this again.

It sounds like @cpuguy83 might be experimenting with something with a more medium-term time horizon? 

If that is slightly further out in time, is there also a near-term improvement that could happen here?

For example, making this up, maybe it might take a N months to decide and implement how Docker might apply semver release tags routinely and in an automated fashion, and maybe it might take M months to figure out how or if Docker might use modules itself... 

But maybe in advance of either of those happening:

1. Could the documentation be updated to at least include a short statement of how a modules-based consumer should import docker today, based on how things stand today? (Related: what issue tracks that: is it this issue here? https://github.com/moby/moby/issues/38507#issuecomment-456225148? Something else?)

2. I don't know enough about how the repositories have been re-arranged to know if this is sensible, but could at least one semver tag be applied on a recent commit in the major Docker repositories, even if done manually, and even if for now it is done on a one-time basis? Currently, modules default to a very old tag for Docker, which in turn then often causes incompatibilities, and that old Docker release also happens to import the old & currently wrong uppercase version of `Sirupsen/logrus`, which in turn causes additional conflicts. Even a one-time tagging with semver tag such as `v1.4.0` to some recent commit would help people stop getting a ~2.5 year old release by default. (And if hypothetically at the end of 2019 it still defaults to a mid-2019 release because bigger changes to process haven't happen yet, then a mid-2019 release at that point is still much better than an early 2017 release).

Here is example of trying to get the latest docker release with modules:
```
$ cd $(mktemp -d)
$ go mod init tempmod
$ go get -d github.com/docker/docker@latest
$ grep docker go.mod
```

which finds a release from early 2017:
```
require github.com/docker/docker v1.13.1 // indirect
```
If interested, see https://github.com/golang/go/issues/28489 for many examples of problems resulting from that old docker tag still being the most recent semver tag, which then in turn often causes `Sirupsen/logrus` vs. `sirupsen/logrus` mismatches, which in turn causes everything to fail.

**Comment by cpuguy83** [comment]
@thepudds There is no way currently to make this repo work nicely  with modules (with version detection and such) and the development of the repo does not fit into semver at all.

Additionally we do not currently have releases on this repo, though we have been been discussing bring that back in some way (and eliminating the docker/engine repo), however this doesn't really fix anything for modules since you'll still need to specify a custom version.

I should point out that the API is backwards compatible with all older clients, however if you need newer functionality you will of course need to update the version of the client.

> Could the documentation be updated to at least include a short statement of how a modules-based consumer should import docker today, based on how things stand today? (Related: what issue tracks that: is it this issue here? #38507 (comment)? Something else?)

I think that's a great idea. I'm thinking this would be good both in godoc form in the client package as well as in the README.md.

**Comment by stevegore** [comment]
Just pointing out that the [installation instructions](https://docs.docker.com/develop/sdk/#install-the-sdks) for the SDK still don't point out how to install a more recent version than 1.13. As such, all of the examples [here](https://docs.docker.com/develop/sdk/examples/) fail miserably.

**Comment by EricHripko** [comment]
I've tried to get Docker code to work with `go mod` to no avail. Ended up giving up and writing a small script that translates `vendor.conf` files from Docker CLI & Engine to `Gopkg.toml`. This gives `dep` enough of a kick to figure out the dependencies and not drown in Docker releases. I've described the journey taken & the method on [my blog](https://eric.hripko.com/2020/02/23/mission-impossible-importing-docker-code-in-go/#last-ditch-attempt) and also am attaching the [Gopkg.toml](https://gist.github.com/EricHripko/5e116b6d533b02ca19c08bff7c389db2) that worked for me to include __19.03.06__ version of the code (current latest release).

**Comment by cpuguy83** [comment]
It seems to be working for me, but I have to specify the version I want. Since we now tagging releases in this repo again I can just specify the tag I want directly in the go.mod without having to use a `replace` statement.

**Comment by cpuguy83** [comment]
I've also been looking at multi-module repos which is a possibility but a bit painful to bootstrap.
In this scenario we'd version the client package as it's own module, the repo would need tags like `client/v<semver>`.

**Comment by cdreier** [comment]
is there a working way to `go get` the latest version with go modules?

i tried pretty much everything and the best i could achieve was v17.12.... even with replace and explicit versions, go mod just overwrites it

**Comment by cpuguy83** [comment]
Go should not be replacing it, but it may have a pseudo-version that contains 17.12.

**Comment by cdreier** [comment]
hmm, if i run `go get github.com/docker/docker/client@v19.03.9` i get `v17.12.0-ce-rc1.0.20200229013735-71373c6105e3+incompatible` installed instead

and there is still no `client.NewClientWithOpts` function in the sdk, so i would need to use the deprecated functions 

```
> go get github.com/docker/docker/client@v19.03.9
go: found github.com/docker/docker/client in github.com/docker/docker v17.12.0-ce-rc1.0.20200514230353-811a247d06e8+incompatible
go: finding module for package github.com/docker/distribution/reference
go: finding module for package github.com/docker/go-units
go: finding module for package github.com/opencontainers/go-digest
go: finding module for package github.com/pkg/errors
go: finding module for package github.com/docker/go-connections/nat
go: finding module for package github.com/docker/go-connections/tlsconfig
go: finding module for package github.com/gogo/protobuf/proto
go: finding module for package github.com/docker/distribution/registry/api/errcode
go: finding module for package github.com/sirupsen/logrus
go: finding module for package github.com/containerd/containerd/errdefs
go: finding module for package google.golang.org/grpc/status
go: finding module for package github.com/docker/go-connections/sockets
go: finding module for package github.com/opencontainers/image-spec/specs-go/v1
go: finding module for package google.golang.org/grpc/codes
go: found github.com/docker/distribution/reference in github.com/docker/distribution v2.7.1+incompatible
go: found github.com/docker/go-connections/sockets in github.com/docker/go-connections v0.4.0
go: found github.com/opencontainers/go-digest in github.com/opencontainers/go-digest v1.0.0
go: found github.com/pkg/errors in github.com/pkg/errors v0.9.1
go: found github.com/docker/go-units in github.com/docker/go-units v0.4.0
go: found github.com/opencontainers/image-spec/specs-go/v1 in github.com/opencontainers/image-spec v1.0.1
go: found github.com/containerd/containerd/errdefs in github.com/containerd/containerd v1.3.4
go: found github.com/sirupsen/logrus in github.com/sirupsen/logrus v1.6.0
go: found google.golang.org/grpc/codes in google.golang.org/grpc v1.29.1
go: found github.com/gogo/protobuf/proto in github.com/gogo/protobuf v1.3.1
go: finding module for package github.com/Microsoft/go-winio
go: found github.com/Microsoft/go-winio in github.com/Microsoft/go-winio v0.4.14
```

any help is appreciated, i think a simple example (like https://docs.docker.com/engine/api/sdk/examples/#list-all-images) on github with go mod and the latest docker client would really help a lot of people trying to start with the docker client sdk

**Comment by thaJeztah** [comment]
> hmm, if i run go get github.com/docker/docker/client@v19.03.9 i get v17.12.0-ce-rc1.0.20200229013735-71373c6105e3+incompatible installed instead

Not sure what it's doing there, perhaps it's because you're trying to get only the `client` package; that commit is two commits "behind" v19.03.8 (https://github.com/moby/moby/compare/71373c6105e3...v19.03.8)

Whereas;

```
go: found github.com/docker/docker/client in github.com/docker/docker v17.12.0-ce-rc1.0.20200514230353-811a247d06e8+incompatible
```

That step shows the correct commit; it's version `v19.03.9`, but go modules do not support release branches, and it generates a "pseudo version" with something it _thinks_ is SemVer. The last part of that pseudo version is the git-commit that's used (`811a247d06e8 `), which matches with the v19.03.0 tag; https://github.com/moby/moby/releases/tag/v19.03.9


Does it work correctly if you `go get github.com/docker/docker@v19.03.9` instead of `/client`?

**Comment by cdreier** [comment]
ohh, this seems to work

```
> go get github.com/docker/docker@v19.03.9
go: github.com/docker/docker v19.03.9 => v17.12.0-ce-rc1.0.20200514230353-811a247d06e8+incompatible
```

after testing with the "list images" example (and fixing the imports) i have a working environment

after the imports the go.mod file expanded to the following, looks pretty similar to the file before, but it builds and is working

```
require (
	github.com/containerd/containerd v1.3.4 // indirect
	github.com/docker/distribution v2.7.1+incompatible // indirect
	github.com/docker/docker v17.12.0-ce-rc1.0.20200514230353-811a247d06e8+incompatible
	github.com/docker/go-connections v0.4.0 // indirect
	github.com/docker/go-units v0.4.0 // indirect
	github.com/gogo/protobuf v1.3.1 // indirect
	github.com/moby/moby v1.13.1
	github.com/opencontainers/go-digest v1.0.0 // indirect
	github.com/opencontainers/image-spec v1.0.1 // indirect
	github.com/pkg/errors v0.9.1 // indirect
	github.com/sirupsen/logrus v1.6.0 // indirect
	google.golang.org/grpc v1.29.1 // indirect
)
```

i test a bit more and create an example repo - thanks a lot for your help!

**Comment by thaJeztah** [comment]
That's good to hear!

I'm still confused what magic go modules apply in your `go get github.com/docker/docker/client..` case (why it picks that random commit)

**Comment by cdreier** [comment]
next step: i found, restarting the go language server also fixed the issue when go getting the explicit client package - both go.mod files looks now the same

```
module test2

go 1.14

require (
	github.com/Microsoft/go-winio v0.4.14 // indirect
	github.com/containerd/containerd v1.3.4 // indirect
	github.com/docker/distribution v2.7.1+incompatible // indirect
	github.com/docker/docker v17.12.0-ce-rc1.0.20200514230353-811a247d06e8+incompatible
	github.com/docker/go-connections v0.4.0 // indirect
	github.com/docker/go-units v0.4.0 // indirect
	github.com/gogo/protobuf v1.3.1 // indirect
	github.com/opencontainers/go-digest v1.0.0 // indirect
	github.com/opencontainers/image-spec v1.0.1 // indirect
	github.com/pkg/errors v0.9.1 // indirect
	github.com/sirupsen/logrus v1.6.0 // indirect
	google.golang.org/grpc v1.29.1 // indirect
)
```

at least the examples are working now, and i can reproduce a working dev environment :) 

as far as i can see, and following the client package, i get https://pkg.go.dev/github.com/docker/docker@v17.12.0-ce-rc1.0.20200514230353-811a247d06e8+incompatible/client?tab=doc

released on may 14th, so perhaps this is just a naming issue? still not understanding why go mod shows v17.12...

**Comment by thaJeztah** [comment]
> still not understanding why go mod shows v17.12...

Go modules look at the tag/version you specified. If that tag is SemVer, it "just works" (picks the version and is done), however, if the tag is non-SemVer, it's treats it as a "non-released version", and starts to construct a pseudo-version that is "valid SemVer". I _think_ earlier versions always used `v0.0.0.0-<date of commit>-<commit-sha>`. Later versions started to be "helpful" and attempt to give a bit of a "clue" what "version" it relates to (so, if you would specify a commit that's 5 commits "ahead" of `v1.1.0`, it would show `v1.1.0-<date>-<commit>`, and if it is some commits "ahead" of `v2.0.0`, it would show `v2.0.0-<date>-<commit>`). Basically the "go modules" equivalent to `git describe`


```bash
$ git describe v19.03.9
v19.03.9
```

But go modules only looks at tags that can be parsed as SemVer (`git describe --match` uses globbing, not regular expressions, so approximating the matching here);

```bash
$ git describe --match 'v[1-9]*.[1-9]*.[0-9]*' v19.03.9
v17.12.0-ce-rc1-3514-g811a247d06
```

It then takes the version it found (`v17.12.0-ce-rc1`), appends the commit date of the version you specified (v19.03.9);

```bash
$ TZ=UTC git show --quiet --date=local --format="%cd" v19.03.9

Thu May 14 23:03:53 2020 # (= 20200514230353 in YYYYMMDDHHmmss)
```

And finally appends the commit matching v19.03.9 (`811a247d06`) and a `+incompatible` (because it think its SemVer `v17.x.x`, so import paths should have `/v17` in them);

```
v17.12.0-ce-rc1.0.20200514230353-811a247d06e8+incompatible
--------------- - -------------  ------------ ------------
       |        |      |              |           |
       |        |      |              |           +-------> "incompatible" because missing "/v17" in import paths
       |        |      |              +-------------------> commit matching v19.03.9
       |        |      +----------------------------------> commit date (in UTC) matching v19.03.9
       |        +-----------------------------------------> 0 added to make it "valid" SemVer
       +--------------------------------------------------> last release / tag that looked like SemVer
```

**Comment by cdreier** [comment]
thanks a lot for your time explaining it in such detail. really helpful!

**Comment by g00nix** [comment]
This ticket has one year of age now and I just wanted to say `Happy Birthday!` to everybody!

I really hope that one day I will be able to copy paste an example from the official documentation, to build and to admire a fully working SDK, without having to do any extra tricks. Of course, it should work with the `latest` libs, not with libs from 2017.

**Comment by cpuguy83** [comment]
There are no tricks you have to do except set the version you want.
Automatic version selection is magic... just set what you want.

**Comment by akerl** [comment]
That’s technically true but dodges expected behavior. Automatic version selection selects the latest tag found. Because of how Docker has approached tags historically vs present, this search for the latest tag finds an ancient tag. This causes Docker imports to behave in an unexpected way, relative to the vast majority of go codebases.

**Comment by cpuguy83** [comment]
@akerl 

1. Not the vast majority of code bases, certainly not ones that have a history
2. go modules is relatively new, and forces a certain versioning scheme that does not make sense for this project. For that matter semver is often incorrectly applied due to ppl not understanding some particular nuance.
3. We cannot change existing tags in this repo in order to fix version selection.
4. It is absolutely trivial to set the version you want
5. If you want to see a change, then propose one, please. I've started a new Docker client project designed to work with semver semantics... but I'm not going to work on it on my own. You are welcome to send PR's and use it for your needs. https://github.com/cpuguy83/go-docker/

**Comment by g00nix** [comment]
>     4\. It is absolutely trivial to set the version you want

No, it is not. Even if you go get a specific version or overwrite go.mod or imports, you will still get older versions for the other packages that get imported. Even if you "get it to work" meaning you can list the containers you have (the most basic example ever), once you start getting into deeper tasks, the code will crash because of incompatibility between libraries that changed over time.

It is not even clear if we should use the `vendor.conf`. I tried that too. That also failed.

At this point, the SDK is still not usable for most people. It is easier to call the `docker` command from bash than to use the libraries, which is a bad sign. If you want to use a SDK, you are better off using the python SDK for docker.


> go modules is relatively new, and forces a certain versioning scheme that does not make sense for this project. For that matter semver is often incorrectly applied due to ppl not understanding some particular nuance.
> We cannot change existing tags in this repo in order to fix version selection.

This mostly sounds like the problem can not be fixed and the ticket should be closed. If that is the case, I guess all we can do is continue to use python or bash.

**Comment by g00nix** [comment]
I tried. That worked. I am fascinated.

I organized the fix a bit for people to be able to test it out:

## fixing with go mod tidy
Copy/paste the go first example from: https://docs.docker.com/engine/api/sdk/#sdk-and-api-quickstart

If you run `go mod tidy` at this point will give you something like this:
```
extreme% cat go.mod 
module go_test

go 1.14

require (
	github.com/Microsoft/go-winio v0.4.14 // indirect
	github.com/docker/distribution v2.7.1+incompatible // indirect
	github.com/docker/docker v1.13.1
	github.com/docker/go-connections v0.4.0 // indirect
	github.com/docker/go-units v0.4.0 // indirect
	github.com/opencontainers/go-digest v1.0.0 // indirect
	github.com/pkg/errors v0.9.1 // indirect
	github.com/stretchr/testify v1.6.1 // indirect
	golang.org/x/net v0.0.0-20200602114024-627f9648deb9 // indirect
)
extreme% go run .
# go_test
./main.go:16:14: undefined: client.NewClientWithOpts
./main.go:16:39: undefined: client.FromEnv
./main.go:16:55: undefined: client.WithAPIVersionNegotiation
./main.go:39:53: undefined: container.WaitConditionNotRunning
extreme% 
```

Replace your `go.mod` with:
```
extreme% cat go.mod 
module go_test

go 1.14

require (
	github.com/docker/docker v19.03.11
)
```
At this point, ignore the error from your IDE and just run `go mod tidy` again. This time your `go.mod` will look totally different:
```
module go_test

go 1.14

require (
	github.com/Azure/go-ansiterm v0.0.0-20170929234023-d6e3b3328b78 // indirect
	github.com/Microsoft/go-winio v0.4.14 // indirect
	github.com/containerd/containerd v1.3.4 // indirect
	github.com/docker/distribution v2.7.1+incompatible // indirect
	github.com/docker/docker v17.12.0-ce-rc1.0.20200531234253-77e06fda0c94+incompatible
	github.com/docker/go-connections v0.4.0 // indirect
	github.com/docker/go-units v0.4.0 // indirect
	github.com/gogo/protobuf v1.3.1 // indirect
	github.com/google/go-cmp v0.4.1 // indirect
	github.com/gorilla/mux v1.7.4 // indirect
	github.com/morikuni/aec v1.0.0 // indirect
	github.com/opencontainers/go-digest v1.0.0 // indirect
	github.com/opencontainers/image-spec v1.0.1 // indirect
	github.com/pkg/errors v0.9.1 // indirect
	github.com/sirupsen/logrus v1.6.0 // indirect
	golang.org/x/sys v0.0.0-20200602225109-6fdc65e7d980 // indirect
	golang.org/x/time v0.0.0-20200416051211-89c76fbcd5d1 // indirect
	google.golang.org/grpc v1.29.1 // indirect
	gotest.tools v2.2.0+incompatible // indirect
)
```

You will be able to run the hello world example:
```
% go run .
{"status":"Pulling from library/alpine","id":"latest"}
{"status":"Digest: sha256:185518070891758909c9f839cf4ca393ee977ac378609f700f60a771a2dfe321"}
{"status":"Status: Image is up to date for alpine:latest"}
hello world

% docker ps -a
CONTAINER ID        IMAGE               COMMAND                CREATED             STATUS                      PORTS               NAMES
b5fd82fa4b3c        alpine              "echo 'hello world'"   11 seconds ago      Exited (0) 10 seconds ago                       gallant_jones
007406207c2b        alpine              "echo 'hello world'"   4 minutes ago       Exited (0) 4 minutes ago                        gracious_liskov
  % 
```

### conclusion

This fix seems to make the SDK work with the latest packages so we can start coding. If you test this, please let me know if you found any compatibility issues with older libs or similar stuff related to the issue described in this ticket or in the comments.

**Comment by johejo** [comment]
This is my bash snippet to bump `github.com/docker/docker`.

```bash
#!/usr/bin/env bash

version="v19.03.11"

sed -i -e "s|github.com/docker/docker v.*$|github.com/docker/docker ${version}|g" go.mod
go mod tidy
```

**Comment by devinbost** [comment]
None of the suggestions discussed here work for me. I am _completely_ blocked. I have tried every workaround I can find. Nothing works. I reported the issue here: https://github.com/moby/moby/issues/41191

**Comment by yihuaf** [comment]
The steps included here are in the right direction. What you have to do is continue to explicitly `require` the module that causes the conflict to a more recent version. For example, I ran into this issue this morning using the same `go.mod` and got:
```
go: kontain.app/km/tools/faktory/image imports
        github.com/docker/docker/image imports
        github.com/docker/docker/layer imports
        github.com/docker/docker/daemon/graphdriver imports
        github.com/docker/docker/plugin/v2 imports
        github.com/docker/docker/oci imports
        github.com/opencontainers/runc/libcontainer/configs imports
        github.com/Sirupsen/logrus: github.com/Sirupsen/logrus@v1.6.0: parsing go.mod:
        module declares its path as: github.com/sirupsen/logrus
                but was required as: github.com/Sirupsen/logrus
```
Go mod is automatically looking at an older version of `runc` release, so I had to explicitly add
```
require (
	github.com/docker/docker v17.12.0-ce-rc1.0.20200531234253-77e06fda0c94+incompatible
	github.com/opencontainers/runc v1.0.0-rc91 //indirect
)
```
And then it seems to all work fine. Format of the version is explained in the comments above in this thread: https://github.com/moby/moby/issues/39302#issuecomment-635253855

**Comment by devinbost** [comment]
As I mentioned in the other issue, this still doesn't work for me.

**Comment by the-maldridge** [comment]
>Note2: I am curious if there is a better solution out there. Or an outline of the efforts needed to make using SDK a bit more friendly.

As others have commented, the solution would be to tag versions that match a well defined semver format so that this 'just works'.  While docker may be standing on a hill of technical righteousness and Go is doing something dumb with versions, this very clearly violates the principal of least surprise.  Having to go muck about with explicit `require` dependencies because of nested dependencies in an upstream project which refuses to integrate with the module versioning system is a very frustrating experience.

I recognize that moby/docker tag binary releases and not API changes.  While this comes off to me as a very strange way to do things, and I'm sure there's a good historical case for it  Unfortunately its also a very compelling argument for me to not go anywhere near the SDK and just call the docker tools from an `os.Exec`, because I can be fairly confident that *that* will work across many different versions.

TL;DR: Go is imposing things across the Go universe that don't make sense in all cases, while moby really appears to be shipping an SDK with zero stability promises and then acting surprised when people have a hard time using it.

**Comment by cpuguy83** [comment]
> While docker may be standing on a hill of technical righteousness and Go is doing something dumb with versions

No one is claiming either of these things.

>  Having to go muck about with explicit require dependencies because of nested dependencies in an upstream project which refuses to integrate with the module versioning system is a very frustrating experience.

No one is refusing to do anything.
We need to come up with a plan that works without breaking everything because it's not just as simple as making a semver tag, because actually some of our tags are technically semver compliant, which is why 17.12 is picked automatically by go.

It is a *minor* inconvenience to set the module version to use that matches the functionality you need. In practice, Docker has not had a major version release in over a year.
While the client SDK API may change between major versions (sadly it is not well designed to not have breaking changes when new features are added), it is stable within that version... so the client at 19.03.0 and 19.03.13 are the same API.

**Comment by cpuguy83** [comment]
By the way, contributions welcome on https://github.com/cpuguy83/go-docker which attempts to resolve these issues.... and is technically something that *could* be brought into moby/moby as a submodule once complete.

**Comment by the-maldridge** [comment]
I'll take a look when I have time.  Right now I'm ~5 yaks deep trying to get linuxkit to build with go modules and then subsequently get a terraform provider to work with that.

**Comment by g00nix** [comment]
> We need to come up with a plan that works without breaking everything because it's not just as simple as making a semver tag, because actually some of our tags are technically semver compliant, which is why 17.12 is picked automatically by go.

Create a new git project with semver. :laughing:

**Comment by cpuguy83** [comment]
@g00nix I did, https://github.com/cpuguy83/go-docker

**Comment by g00nix** [comment]
> @g00nix I did, https://github.com/cpuguy83/go-docker

I was actually serious. Changing the version system of docker to semver will cause issues with all distros that have docker as a package. This will not be fun.

If people think semver is actually required, they could create a new official project with semver and the package managers will migrate from the old package to the new one. I don't see any other smooth way to do it.

There always is the alternative of accepting the idea that the SDK is mostly dead and that people will keep using `os.Exec`, since changing project name and versioning system is too much effort for the entire community.

**Comment by cpuguy83** [comment]
@the-maldridge Feel free to ping me on the docker community slack if you need some help.
There should not be any reason that a repo needs to adopt go modules for its dependency management in order to use it with a downstream project that does use modules.

@g00nix I'm serious too, see the readme in that repo. Go does support submodules and we can absolutely look at doing something there.

**Comment by thepudds** [comment]
Hi  @cpuguy83 

> which is why 17.12 is picked automatically by go.

When does that happen?  AFAICT, the 17* tags should not be picked by default by cmd/go, but v1.13.1 tag will. 

(Perhaps you are not speaking of tags? Or perhaps I misunderstood ;-)

**Comment by cpuguy83** [comment]
@thepudds Probably you already had 1.13.1 cached before we added tags back to the repo. 17.12.x is the latest semver compatible tag. If we can manage to get docker 20 out the door, that will be the new latest semver compatible tag.

**Comment by thepudds** [comment]
Hi @cpuguy83, I’m not sure that is correct. 

For why I think that, see my write up in the section:

  _A few more details about "Suggestion 1: apply a v1.20.x tag or similar"._

in https://github.com/moby/moby/issues/41503#issue-711280816

**Comment by neersighted** [comment]
Closing this for now as moby/moby is now the canonical upstream repository, and we have migrated to semver-compatible versioning as part of the plan to transition to be a Go module.

**Comment by t2y** [comment]
@neersighted I'm waiting for supporting Go modules by Moby, so I want to subscribe to an issue to know it. Could you tell me the issue to subscribe?

**Comment by neersighted** [comment]
There isn't one yet, but I was talking to @crazy-max about creating a tracking issue and making it a goal for the 26.0 release. Would you mind linking back here once you create such an issue?

**Comment by neersighted** [comment]
https://github.com/moby/moby/pull/46772 is likely relevant for those watching this issue.
